components/hab/src/cli.rs

```diff
@@ -1209,8 +1209,8 @@ fn sub_svc_load() -> App<'static, 'static> {
             "Application name; [default: not set].")
         (@arg ENVIRONMENT: --environment -e +takes_value requires[APPLICATION]
             "Environment name; [default: not set].")
-        (@arg CHANNEL: --channel +takes_value default_value[stable]
-            "Receive package updates from the specified release channel")
+        (@arg CHANNEL: --channel +takes_value
+            "Receive package updates from the specified release channel [default: stable]")
```

Comment:
Since we've got a type that encapsulates both the default value and the environment variable the data responds to, we can take advantage of that in our CLI definition.

```rust
(@arg CHANNEL: --channel +takes_value env(ChannelIdent::ENVVAR) default_value(ChannelIdent::default_as_str())
 "Receive package updates from the specified release channel")
```

This also requires adding this over in `components/core/src/lib.rs`:
```rust
default_as_str!(ChannelIdent);
```
That last macro is something we wrote to help bridge the existing gap in functionality of Clap to expose such typed default values in the CLI.

The resulting help text looks like this:
```sh
...
--channel <CHANNEL>
     Receive package updates from the specified release channel [env: HAB_BLDR_CHANNEL=]  [default: stable]
...
```
Or this, if you have `HAB_BLDR_CHANNEL` set:
```sh
...
--channel <CHANNEL>
     Receive package updates from the specified release channel [env: HAB_BLDR_CHANNEL=foo]  [default: stable]
...
```
(Since this could respond to an environment variable, it's important to document that fact.)

(Also note that this pattern isn't uniformly followed throughout the CLI, as it's only been recently "discovered". If you see instances where defaults are hardcoded in documentation strings, then that's an indicator that some more refactoring could take place.)

Reply:
One thing that I'm not quite sure the best way to address is the fact that when you start up a Supervisor, it has a `--channel` and `--url` option that apply to where the Supervisor pulls updates from, if it is auto-updating. That channel _does not_ respond to `HAB_BLDR_CHANNEL` (I'm not sure if that was intentional, or not), but the URL _does_ respond to `HAB_BLDR_URL`. 

This PR would add a "hidden" meaning for `HAB_BLDR_CHANNEL`, and a _double_ meaning for `HAB_BLDR_URL`. The overall layering and precedence of these, as you laid out in the PR message, make sense, and provide ample opportunity for overriding, and the modified log message you include also provides visibility, which is fantastic. However, it seems like there's perhaps room for additional feedback or further documentation / clarification about how these environment variables interact, and which portions of code are sensitive to them and which aren't.

This also brings to mind #7033... I wonder if that might be a more appropriate place to define these kind of "global default" values that makes it easier to reason about? (genuinely asking; I'm not sure what the final answer should be just yet).

@krasnow, @apriofrost: do you have any thoughts on this?

Reply:
> Since we've got a type that encapsulates both the default value and the environment variable the data responds to, we can take advantage of that in our CLI definition.
> 
> ```rust
> (@arg CHANNEL: --channel +takes_value env(ChannelIdent::ENVVAR) default_value(ChannelIdent::default_as_str())
>  "Receive package updates from the specified release channel")
> ```

@christophermaier I really appreciate that you pointed out the `env()` clap macro! 🎉  That definitely helps provide more UX context when looking at the help text. I've incorporated that in this [commit](https://github.com/habitat-sh/habitat/pull/7215/commits/382bd6a1ac2f3519f38e1afd8c2672ce91a59d27) for both BLDR_URL and CHANNEL args.

I did omit the `default_value(ChannelIdent::default_as_str()` though considering I wanted the defaults to be applied from the Supervisor context, when it instantiates the ServiceSpec. If we omit the default at the client cli level, it will pass [`Option<None>`](https://github.com/habitat-sh/habitat/blob/master/components/hab/src/main.rs#L1959-L1960) for both in the ctl_message. The `None` for both won't then override the defaults when the Service Manager [merges](https://github.com/habitat-sh/habitat/blob/master/components/sup/src/manager/service/spec.rs#L242-L247) in the values passed via the ctl_message. Otherwise, if we set a default at the client context, it will override any defaults from the Supervisor context and we lose the intent of this PR since the client will _always_ pass a `Some` value.

I'm curious what your thoughts are on that idea 😃

Reply:
> One thing that I'm not quite sure the best way to address is the fact that when you start up a Supervisor, it has a `--channel` and `--url` option that apply to where the Supervisor pulls updates from, if it is auto-updating. That channel _does not_ respond to `HAB_BLDR_CHANNEL` (I'm not sure if that was intentional, or not), but the URL _does_ respond to `HAB_BLDR_URL`.
> 
> This PR would add a "hidden" meaning for `HAB_BLDR_CHANNEL`, and a _double_ meaning for `HAB_BLDR_URL`.

My instincts are telling me that what we need is consistency to make this less confusing and more predictable. If we do these two things for `hab sup run`, I think we keep true to that ideal:

* make `CHANNEL` respond to `HAB_BLDR_CHANNEL`
* use the clap `env()` macro for _both_ `BLDR_URL` and `CHANNEL` args

The result is that existing `HAB_BDLR_URL` and `HAB_BLDR_CHANNEL` envvars in the Supervisor or cli context act with a global context and apply to _both_ Supervisor and Services. The values from the ENVVARs would be apparent to the user via the help text due to the `env()` clap macro. If the user wishes to set them differently for different contexts, they can be more specific by utilizing the `--url` and `--channel` arguments.

I believe that will still marry up with #7033 if we make the same arguments from the Supervisor config file the _most_ specific with the highest precedence level or we could even create distinguishing configuration for Supervisor vs Service. 
 
 @christophermaier, @krasnow, @apriofrost: thoughts?

Reply:
:thinking: Making a change like this to cause the Builder URL and channel apply to both the updating supervisor _and_ services might not great. One thing that seems a bit weird is that the environment variables would apply to both, but the corresponding CLI options only apply to the Supervisor.

I wonder if it might be a good idea to go ahead and separate those things?

Reply:
> Ah, that's an excellent point! That should most definitely be documented with a comment in the CLI code, then, because if I didn't know that background, that would be one of the first things I would change in this code, thus breaking everything!

Ahh yes, 1000% agree with this! I've added comments in https://github.com/habitat-sh/habitat/pull/7215/commits/66bbfc007deb5ad35d90192cf43b1955e10c14d3 Additionally the (to-be-added) e2e tests will help cover the code, providing some safety.

> 🤔 Making a change like this to cause the Builder URL and channel apply to both the updating supervisor and services might not great. One thing that seems a bit weird is that the environment variables would apply to both, but the corresponding CLI options only apply to the Supervisor.
> I wonder if it might be a good idea to go ahead and separate those things?

Great point! I do see the `HAB_BLDR_URL` and `HAB_BLDR_CHANNEL` CLI combo options for Supervisor and Services are distinct already:

* The Supervisor's options for `hab sup run` are defined [here](https://github.com/habitat-sh/habitat/blob/master/components/hab/src/cli.rs#L1081-L1086) which apply to Supervisor updates.
* The ServiceSpec options from `hab svc load` are defined [here](https://github.com/habitat-sh/habitat/blob/master/components/hab/src/cli.rs#L1212-L1219) which apply to the Service updates.

If we move forward with this idea, it would mean the Environment variables apply to both, either from the cli or the Supervisor context. Distinction can still be asserted by specifying the CLI arguments (which are a higher precedence) as needed for either the Supervisor (`hab sup run`) or the Service (`hab svc load`).

Reply:
@christophermaier I'm curious what your thoughts are on the latest couple of commits and the included tests. I do think the work here solves the user problem from the issue, however I can understand if we want to hold off on it or pivot to a different strategy. Of course, I would just like to move forward in the best way possible for the project. Thanks!